File: conv.clj
(ns conv
  (:require [clojure.string :as s]))

;; /\*.*?\*/ matches annotate
;; #[^\n]+\n matches ignored comment
;; [^\n;{}]+ matches any line
;; ;|\{|} matches ; or { or }

(defn process [s]
  (let [f (fn [{:keys [tags temp] :as r} piece]
            (cond
              (= [\/ \*] (take 2 piece)) (-> r
                                             (update :anno conj "top")
                                             (update :anno conj (s/join " " (cons "edit" tags)))
                                             (assoc :anno-msg (str " \"" (subs piece 3 (- (count piece) 3)) "\"")))
              (= \# (first piece)) r
              (= "{" piece) (-> r (update :tags conj temp) (assoc :temp nil))
              (= "}" piece) (-> r (update :tags pop))
              (= ";" piece) (-> r
                                (update :lines conj (s/join " " (cons "set" (conj tags temp))))
                                (assoc :temp nil))
              (= "inactive:" (apply str (take 9 piece))) (-> r
                                                             (update :lines conj (s/join " " (cons "deactivate" (conj tags (subs piece 10)))))
                                                             (assoc :temp (subs piece 10)))
              (= "protect:" (apply str (take 8 piece))) (-> r
                                                            (update :lines conj (s/join " " (cons "protect" (conj tags (subs piece 9)))))
                                                            (assoc :temp (subs piece 9)))
              :else (cond-> r
                      (:anno-msg r) (update :anno conj (str "annotate " piece (:anno-msg r)))
                      :then (dissoc :anno-msg)
                      :then (update :temp #(if % (str % " " piece) piece)))))
        {:keys [lines anno]} (->> (re-seq #"/\*.*?\*/|#[^\n]+\n|[^\n;{}]+|;|\{|}" s)
                                  (map s/trim)
                                  (remove #{""})
                                  (reduce f {:tags [], :anno [], :lines [], :temp nil}))]
    (->> (lazy-cat lines anno) (s/join "\n"))))

(if (= 1 (count *command-line-args*))
  (->> (first *command-line-args*) slurp process println)
  (println "Usage: clj conv.clj <filename>"))

File: conv.py
import re
import sys

# /\*.*?\*/ matches annotate
# #[^\n]+\n matches ignored comment
# [^\n;{}]+ matches any line
# ;|\{|} matches ; or { or }

def process(data):
    lconfig = []
    lannotations = []
    annotation = ''
    ltags = []
    temp = ''

    for elem in re.findall(r"/\*.*?\*/|#[^\n]+\n|[^\n;{}]+|;|\{|}", data):
        elem = elem.strip()

        if elem == '' or elem[0] == '#':
            continue
        elif elem[0:2] == '/*':
            lannotations.append("top")
            lannotations.append("edit " + " ".join(ltags))
            annotation = '"' + elem[3:-3] + '"'
        elif elem[0] == '{':
            ltags.append(temp)
            temp = ''
        elif elem[0] == '}':
            ltags.pop()
        elif elem[0] == ';':
            lconfig.append("set "        + " ".join(ltags) + " " + temp)
            temp = ''
        elif elem[0:9] == 'inactive:':
            lconfig.append("deactivate " + " ".join(ltags) + " " + elem[10:])
            temp = elem[10:]
        elif elem[0:8] == 'protect:':
            lconfig.append("protect "    + " ".join(ltags) + " " + elem[9:])
            temp = elem[9:]
        else:
            if annotation:
                lannotations.append("annotate " + elem + " " + annotation)
                annotation = ''
            if temp:
                temp = temp + " " + elem
            else:
                temp = elem

    for a in lconfig:
        print(a)
    for a in lannotations:
        print(a)

if len(sys.argv) != 2:
    print("Usage: python conv.py <filename>")
else:
    process(open(sys.argv[1], 'r').read()
